Resource economy + UI

TToJlkoBHuK opened the pull request:
1) Добавлена экономика начисления ресурсов
2) Добавлена кнопка начисления ресурсов
3) Добавлено визуальное отображение начисляемых ресурсов
4) Добавлен uasset конкретного ресурса, который интегрирован в общий UI

Update:
1) изменен UI Widget путем добавления 2 и 3 пункта
2) изменено строительство дома на шахту, это было для проверок, можно вернуть обратно
3) обновился какой-то дефолт эдитор, я не понял что это.

как примерно работает:
когда нажимаем кнопку resource - запускается Economy компонент, который запускает цикл проверки зданий на сетке, то есть он смотрит, что находится в ячейке, если он видит что это здание, если это ресурсное здание, смотрит, какой у него добываемый тип и сколько и записывает в общую кучу всех ресурсов. Как только сетка пройдена - начисление всех ресурсов согласно типу и количеству*количество зданий. Действует ограничение по максимальному, после не начисляет просто. визуальное отображение появляется динамически, пока не придумал, как статично держать, если придумаете, будет хорошо. На скриншотах UI вылезает, но это уже пофикшено до нормальных размеров.

<img width="1919" height="1079" alt="Снимок экрана 2025-12-22 005134" src="https://github.com/user-attachments/assets/492b7a2b-b35d-468b-92f6-2f6e0f95d090" />
<img width="1919" height="954" alt="Снимок экрана 2025-12-22 005151" src="https://github.com/user-attachments/assets/05a314a3-83eb-432e-b75d-aa06e9b7c309" />

TToJlkoBHuK committed:
update ResourceType #123

add food, population, progress

TToJlkoBHuK committed:
add Economy component #123

"Resource Collector": It scans the grid of buildings, finds all resource buildings, and immediately adds the resources they produce to the player's bank.

TToJlkoBHuK committed:
add UI Resource icon widget #122

UI of a specific resource: it is responsible for the visual display of icons and text (for example, "50/100") for a single line in the interface.

TToJlkoBHuK committed:
update Resource for UI #122

TToJlkoBHuK committed:
UI Resource #122

TToJlkoBHuK committed:
update Widget #122

add Resource button
add UI Resource

TToJlkoBHuK committed:
update #122

честно не понял, что за файл у меня обновился, к чему он относится, но его надо закоммитить.

Makintosh365 commented on Lords_Frontiers/Source/Lords_Frontiers/Private/ResourceManager/EconomyComponent.cpp:
```diff
@@ -0,0 +1,99 @@
+    TMap<EResourceType, int32> CollectedTotals;
+    const int32 Width = GridManager_->GetGridWidth();
+    const int32 Height = GridManager_->GetGridHeight();
+
```
Лучше всего будет сделать обход не по сетке это может быть очень дорого
Наверно будет правильнее что то такое  TArray<TWeakObjectPtr<AResourceBuilding>> ResourceBuildings;

Makintosh365 commented on Lords_Frontiers/Source/Lords_Frontiers/Private/ResourceManager/EconomyComponent.cpp:
```diff
@@ -0,0 +1,99 @@
+                UResourceGenerator* Gen = ResBuilding->GetResourceGenerator();
+                if (!Gen)
+                {
+                    UE_LOG(LogTemp, Error, TEXT("Building %s has NO ResourceGenerator component!"), *Actor->GetName());
```
Как выше

Makintosh365 commented on Lords_Frontiers/Source/Lords_Frontiers/Private/ResourceManager/EconomyComponent.cpp:
```diff
@@ -0,0 +1,99 @@
+
+void UEconomyComponent::CollectGlobalResources()
+{
+    if (!GridManager_ || !ResourceManager_)
```
Весь if похож на костыль )

Makintosh365 commented on Lords_Frontiers/Source/Lords_Frontiers/Private/ResourceManager/ResourceItemWidget.cpp:
```diff
@@ -0,0 +1,19 @@
+}
+
+void UResourceItemWidget::UpdateAmount(int32 NewAmount, int32 MaxAmount)
+{
```
Мб стоит сохранять текущее значение на будущие. Не критично. Возможно так будет проще потом работать с "анимациями" чисел 
А так все норм

Makintosh365 commented on Lords_Frontiers/Source/Lords_Frontiers/Public/ResourceManager/GameResource.h:
Можно добавить MAX UMETA(Hidden) для итерации и подсчёта
в самый конец

TToJlkoBHuK committed:
Implementation of automatic registration of resource buildings in the economy system #122

1) В класс AResourceBuilding добавлен метод EndPlay для корректной очистки ссылок.
2) Реализована автоматическая регистрация здания в UEconomyComponent при вызове BeginPlay.
3) Реализована автоматическая отписка (unregistration) здания при его разрушении или завершении игры в EndPlay.

TToJlkoBHuK committed:
Eliminating magic numbers and improving code reuse

1) В заголовочный файл добавлена константа cDefaultMaxResource, заменяющая жестко прописанное значение "100"
2) Метод AddResource обновлен для использования существующего геттера GetMaxResourceAmount вместо дублирования логики проверки лимитов.
3) Метод GetMaxResourceAmount теперь возвращает именованную константу вместо «магического числа».

TToJlkoBHuK committed:
Technical improvement of resource enumerations and extension of widget status #122

1) В перечисление EResourceType добавлен маркер MAX (UMETA Hidden)
2) В класс UResourceItemWidget добавлена переменная CurrentValue_ для кэширования текущего значения ресурса внутри виджета

TToJlkoBHuK committed:
Transition from grid scanning to building registration system #122

1) Полностью удалена зависимость от GridManager в логике сбора ресурсов. Теперь вместо перебора всех клеток сетки (O(N×M)) используется прямой список активных зданий (O(K))
2) Добавлены методы RegisterBuilding и UnregisterBuilding для динамического отслеживания ресурсных зданий.
3) Реализован механизм «ленивого поиска» зданий через GetAllActorsOfClass, который срабатывает как fail-safe, если список регистраций пуст. Это гарантирует подхват зданий, созданных до инициализации компонента.
4) Для хранения зданий используется TArray с TWeakObjectPtr, что предотвращает утечки памяти и ошибки при обращении к уже удаленным объектам.

TToJlkoBHuK committed:
optimizing connections, memory security, and initialization #122

1) Добавлен вспомогательный метод-хелпер GetResourceManager(), который централизует поиск компонента и заменяет повторяющиеся блоки кода
2) Метод CreateResourceWidgets теперь принимает указатель на UResourceManager напрямую.
3) Все контейнеры TMap (ResourceIcons, ResourceWidgetsMap) переведены на использование TObjectPtr.
4) В OnButton9Clicked добавлена передача ссылки ResourceManager в EconomyComponent через SetResourceManager.
5) Цикл итерации ресурсов переведен на использование EResourceType::MAX
6) Удалены избыточные вызовы UpdateAmount, так как обновление теперь инкапсулировано внутри SetupItem

TToJlkoBHuK committed:
formating code #122

Makintosh365 reviewed (approved):
Сейчас пойдет но нужно будет поправить
Неэффективный поиск зданий GetAllActorsOfClass лучше чтобы здания сами записывались туда и выписывались
Думаю что сливать можно
